### PR TITLE
Column chooser for Messages List window and persistence of the layout

### LINKED ISSUE #734: Add a column chooser
Column selection to be persisted across sessions.

## Plan of attack

- [x] Add column chooser
- [x] Persist across sessions

### PR DESCRIPTION
**Pull request by HEskandari**
Closes #734

### CODE REVIEW

**Comment by HEskandari** [comment]
I have separated the column chooser bits from PR #730. Please review.

File: src/ServiceInsight/ExtensionMethods/LayoutExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+            {
+                using (var ms = new MemoryStream())
+                {
+                    control.SaveLayoutToStream(ms);
+                    return ms.GetAsString();
+                }
+            }
+            catch (Exception ex)
+            {
+                LogTo.Information(ex, "Failed to save the layout, reason is: {ex}", ex);
```
Comment (WilliamBZA) [inline src/ServiceInsight/ExtensionMethods/LayoutExtensions.cs]: Is Info the right log level here? It doesn't cause the app to crash, but perhaps debug is appropriate.

Comment (adamralph) [reply src/ServiceInsight/ExtensionMethods/LayoutExtensions.cs]: Given this is a non-critical failure, I'd say WARN is the appropriate level.

File: src/ServiceInsight/MessageList/MessageListViewModel.cs [hunk 2]
```diff
@@ -286,5 +293,15 @@ public void BringIntoView(StoredMessage msg)
...
+
+        public void OnSavePartLayout()
+        {
+            view.OnSaveLayout(settingsProvider);
+        }
+
+        void RestoreLayout()
+        {
+            view.OnRestoreLayout(settingsProvider);
+        }
```
Comment (adamralph) [inline src/ServiceInsight/MessageList/MessageListViewModel.cs]: I see that `RestoreLayout` is called from `OnViewAttached`. Is there an equivalent method that could call `OnSavePartLayout`, to remove the need for the `IPersistPartLayout` interface, and its explicit call from `ShellViewModel`?

File: src/ServiceInsight/MessageList/MessageListView.xaml.cs [hunk 3]
```diff
@@ -102,5 +106,25 @@ void OnGridClicked(object sender, MouseButtonEventArgs e)
...
+        {
+            var layoutSetting = settingsProvider.GetSettings<MessageListSettings>();
+            grid.RestoreLayout(layoutSetting.GridLayout.GetAsStream());
+        }
+
+        public void OnResetLayout(ISettingsProvider settingsProvider)
+        {
+            var layoutSettings = settingsProvider.GetSettings<MessageListSettings>();
+            layoutSettings.GridLayout = null;
+            settingsProvider.SaveSettings(layoutSettings);
```
Comment (adamralph) [inline src/ServiceInsight/MessageList/MessageListView.xaml.cs]: Why does nothing happen with `grid` in this method?

File: src/ServiceInsight/ExtensionMethods/LayoutExtensions.cs (hunk 1 above)
Comment (adamralph) [reply src/ServiceInsight/ExtensionMethods/LayoutExtensions.cs]: After further review, I see that these log messages are pre-existing and have been moved from `Shell/Shell.xaml.cs`, i.e. the saving and restoration of the new settings for the column chooser are following the pattern of that of the existing settings for the layout of shell. For that reason I think we should keep this as-is for this PR. If we want to change the log level and messages (I think they also need some work) for problems in persisting all settings, I believe we should do that a a separate PR.

File: src/ServiceInsight/Shell/ShellViewModel.cs [hunk 4]
```diff
@@ -132,15 +133,23 @@ protected override void OnViewAttached(object view, object context)
 
         protected override void OnDeactivate(bool close)
         {
-            base.OnDeactivate(close);
             refreshTimer.Stop();
             SaveLayout();
+            base.OnDeactivate(close);
```
[the remark this answers is not in the document]
Comment (HEskandari) [reply src/ServiceInsight/Shell/ShellViewModel.cs]: the base should be called last, as it deactivates the screen and sets a few other flags, not allowing SaveLayout to properly run.

File: src/ServiceInsight/Shell/ShellViewModel.cs [hunk 5]
```diff
@@ -132,15 +133,23 @@ protected override void OnViewAttached(object view, object context)
...
         {
             if (!comandLineArgParser.ParsedOptions.ResetLayout)
             {
+                var persistableLayouts = Items.OfType<IPersistPartLayout>();
+                {
+                    foreach (var screen in persistableLayouts)
+                    {
+                        screen.OnSavePartLayout();
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (HEskandari) [reply src/ServiceInsight/Shell/ShellViewModel.cs]: Here's how it works:

Items is a property of all the parts/sub-screens in the shell. Some of the parts/sub-screens may implement the interface `IPersistPartLayout` that says they need to serialize their layout upon exit. 

So to answer: if we move `Item.OfType<IPersistPartLayout>()` to a well-named method, would it become simpler?

File: src/ServiceInsight/MessageList/MessageListViewModel.cs (hunk 2 above)
Comment (HEskandari) [reply src/ServiceInsight/MessageList/MessageListViewModel.cs]: The ViewAttach stuff comes from `IViewAware` interface of caliburn. It doesn't have anything for the view is detached.

The interface `IPersistPartLayout` still would be required though, as it is an opt-in interface.

File: src/ServiceInsight/MessageList/MessageListView.xaml.cs (hunk 3 above)
Comment (HEskandari) [reply src/ServiceInsight/MessageList/MessageListView.xaml.cs]: This is the Reset method, which would null-out the stored layout in the settings (so it falls back to the default layout). This setting - generally speaking, all settings - become effective on the next restart (we have a UI hint for it as well).

File: src/ServiceInsight/ExtensionMethods/LayoutExtensions.cs (hunk 1 above)
Comment (HEskandari) [reply src/ServiceInsight/ExtensionMethods/LayoutExtensions.cs]: 👍  on doing it separately.

File: src/ServiceInsight/Shell/ShellViewModel.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (adamralph) [reply src/ServiceInsight/Shell/ShellViewModel.cs]: @HEskandari I believe I understand how it works, and the implementation looks fine.

All I'm suggesting is replacing this block:

```C#
var persistableLayouts = Items.OfType<IPersistPartLayout>();
{
    foreach (var screen in persistableLayouts)
    {
        screen.OnSavePartLayout();
    }
}
```

with this block:

```C#
foreach (var item in Items.OfType<IPersistPartLayout>())
{
    item.OnSavePartLayout();
}
```

It's functionally identical, and a pure refactoring. Would that work?

[the remark this answers is not in the document]
Comment (HEskandari) [reply src/ServiceInsight/Shell/ShellViewModel.cs]: @adamralph ah sure :) that works just fine. I'll update.

**Comment by HEskandari** [comment]
@adamralph The PR is update (and rebased). I'll squash the commits after the reviews are done.

**Comment by HEskandari** [comment]
@adamralph all done.

**Pull request merged**